### ISSUE TITLE
`reportReturnType` false positive when attribute type is exhaustively narrowed

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#682]
Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBAJghgFzgYwDZwM4YKYagSwgAcwQFZEV0sBYAKFEigQE8j8A7AcwONPIAy%2BBNhBxU9egAF4SNJgz15WKADEwYAFz0ouqADdxAV2yaoQkWNQBtAIwAaKACYAupLoxswKMA0AKXy01DQBKKABaAD4oACMNVDMAYihRcBAdPXxvQIA6Q1QTKABeIqhbbTo9KqgQbAQjEA4oABUQEwzdbFQsnw0842xi0qcK6r0xfBwoAFEAD2RsIgR8MA4OqHpPbxi4EACNMwtRcTtHVzCo2PikqA4wFJA09Z7AobLRsZq6hqbW9sq9F0Xho3iN1lUJlM5gslisOEA)

```python
from dataclasses import dataclass
from typing import Literal

@dataclass
class Foo:
    value: Literal[1, 2]

def foo(foo: Foo) -> bool: # error
    if foo.value == 1:
        return True
    elif foo.value == 2:
        raise Exception
    
def bar(foo: Literal[1, 2]) -> bool: # no error
    if foo == 1:
        return True
    elif foo == 2:
        raise Exception
```

```
Function with declared return type "bool" must return value on all code paths
  "None" is not assignable to "bool"  (reportReturnType)
```
